csrc/ir/builder.cpp

```diff
@@ -355,9 +355,132 @@ Val* SimplifyingIrBuilder::addExpr(Val* lhs, Val* rhs) {
     return addExpr(rhs, lhs->value());
   } else if (rhs->isConst()) {
     return addExpr(lhs, rhs->value(), rhs->dtype());
-  } else {
+  }
+  // Flatten nested additions. For example, suppose we have
+  //   lhs = a + ( b + c )
+  //   rhs = d + ( ( -b ) + e )
+  // Then we would flatten this to a sum of the terms
+  //   a b c d ( -b ) e
+  // Now suppose we have
+  //   lhs = a + b
+  //   rhs = d + ( -( b + e ) )
+  // We can flatten this to the sum of
+  //   a b d ( -b ) ( -e )
+  // We track a b d as positive terms and b e as negative terms, after which
+  // we can cancel b easily
+  std::vector<Val*> pos_terms;
+  std::vector<Val*> neg_terms;
+
+  // We'll do a stack-based search, using two stacks: one for positive and one
+  // for negative terms. Whenever the negative term stack is non-empty, we
+  // draw from it first.
+  std::vector<Val*> pos_term_stack{lhs, rhs};
+  std::vector<Val*> neg_term_stack;
+  while (!pos_term_stack.empty() || !neg_term_stack.empty()) {
+    Val* term = nullptr;
+    bool term_is_neg = false;
+    if (!neg_term_stack.empty()) {
+      term = neg_term_stack.back();
+      neg_term_stack.pop_back();
+      term_is_neg = true;
+    } else {
+      term = pos_term_stack.back();
+      pos_term_stack.pop_back();
+    }
+    if (auto* uop = dynamic_cast<UnaryOp*>(term->definition());
+        uop && uop->getUnaryOpType() == UnaryOpType::Neg) {
+      Val* unneg = uop->in();
+      if (term_is_neg) {
+        // This is a term like -(-a) or (-b) within -(a + (-b) ), so we treat
+        // it as a positive term
+        pos_term_stack.push_back(unneg);
+      } else {
+        neg_term_stack.push_back(unneg);
+      }
+      continue;
+    } else if (auto* bop = dynamic_cast<BinaryOp*>(term->definition())) {
+      if (bop->getBinaryOpType() == BinaryOpType::Add) {
+        if (term_is_neg) {
+          neg_term_stack.push_back(bop->lhs());
+          neg_term_stack.push_back(bop->rhs());
+        } else {
+          pos_term_stack.push_back(bop->lhs());
+          pos_term_stack.push_back(bop->rhs());
+        }
+        continue;
+      } else if (bop->getBinaryOpType() == BinaryOpType::Sub) {
+        if (term_is_neg) {
+          neg_term_stack.push_back(bop->lhs());
+          pos_term_stack.push_back(bop->rhs());
+        } else {
+          pos_term_stack.push_back(bop->lhs());
+          neg_term_stack.push_back(bop->rhs());
+        }
+        continue;
+      }
+    }
+    if (term_is_neg) {
+      neg_terms.push_back(term);
+    } else {
+      pos_terms.push_back(term);
+    }
+  }
+  // Now do cancellation
+  std::vector<bool> cancelled_pos(pos_terms.size(), false);
+  std::vector<bool> cancelled_neg(neg_terms.size(), false);
+  bool performed_cancellation = false;
+  for (size_t j : c10::irange(neg_terms.size())) {
```

Comment:
Iterate over negative terms first since they are likely more rare. If no negative terms were found, we'll quickly see that no cancellation is possible and return the simple sum.